Add `Pool::add`

tneely opened the pull request:
Fixes #212

This adds `Pool::add`, which allows for externally created connections to be added and managed by the pool. If the pool is at maximum capacity when this method is called, it will return the input connection as part of the Err response.

I considered allowing `Pool:add` to ignore `max_size` when adding to the pool, but felt it could lead to confusion if the pool is allowed to exceed its capacity in this specific case.

This change required making PoolInternals::approvals visible within the crate to get the approval needed to add a new connection. The alternative would have required defining a new pub(crate) method for this specific use case, which feels worse. I'm open to suggestions on how to more cleanly integrate this change into the package.

djc commented on bb8/src/inner.rs:
```diff
@@ -161,6 +161,20 @@ where
     }
 
+    /// Adds an external connection to the pool if there is capacity for it.
+    pub(crate) fn try_put(&self, conn: M::Connection) -> Result<(), M::Connection> {
```
It might be better to move this logic into a method on `SharedPool`, which would also allow us to keep `approvals()` private. Then, I think we want a layer here which would run `manager.has_broken()` on the new `conn` before we add it. Might need another error enum to accomodate the different error paths.

tneely replied:
That makes sense! Let me move things around

djc commented on bb8/src/inner.rs:
```diff
@@ -161,6 +161,20 @@ where
+    pub(crate) fn try_put(&self, conn: M::Connection) -> Result<(), M::Connection> {
+        let mut locked = self.inner.internals.lock();
+        let mut approvals = locked.approvals(&self.inner.statics, 1);
+        match approvals.next() {
```
Suggest spelling this as `let Some(approval) = approvals.next() else { return Err(conn); }`.

tneely replied:
Sounds good!

tneely commented:
Sorry for the delay - I lost track of this PR at some point. I've moved a few things around based on your feedback and also added an `AddError` enum to differentiate between when the connection is broken and when there's no capacity.

djc commented on bb8/src/api.rs:
```diff
@@ -547,3 +555,36 @@ impl<E> ErrorSink<E> for NopErrorSink {
+
+impl<E> fmt::Display for AddError<E>
+where
+    E: error::Error + 'static,
```
Nit: please put these bounds into the `impl` declaration for conciseness (so `impl<E: error::Error + 'static>`).

djc commented on bb8/src/api.rs:
```diff
@@ -547,3 +555,36 @@ impl<E> ErrorSink<E> for NopErrorSink {
 }
+
+/// Error type returned by `Pool::add(conn)`
+#[derive(Debug, Clone, PartialEq, Eq)]
```
Nit: let's move this next to `RunError` (so above `ErrorSink`).

djc reviewed (commented):
Looking pretty good, some small suggestions left!

djc commented on bb8/tests/test.rs:
```diff
@@ -104,6 +104,37 @@ where
+    }
+}
+
+#[async_trait]
```
Sorry, missed this previously: I want to review this move separately, so please isolate it in a separate commit (and squash all the other commits) together.

djc replied:
The idea here is that I want the move of the existing code here isolated in a separate commit from adding the new test -- sorry if that wasn't clear from my previous comment.

tneely replied:
I think I understand? Let me try again.

djc replied:
Not quite, but I've taken the liberty of fixing it up.

djc reviewed (approved):
This mostly looks good, needs one tweak to the commit history.

djc commented:
Thanks for your patience through a number of iterations!

tneely commented:
You're welcome, thanks for working through the PR with me!
